File: 26S/session4.R
# session4.R
# Session 4: Matching Methods & Propensity Scores
# End-to-end runnable example using the Dehejia-Wahba LaLonde-PSID1 sample
# (185 NSW treated + 2,490 PSID-1 observational controls). The severe
# imbalance between these two groups is what makes matching interesting --
# the original NSW experimental sample is already balanced by randomization.
# Demonstrates: PSM, optimal matching, balance diagnostics, inference,
# multiple controls, Mahalanobis hybrid, and DR-style inference.
#
# Required packages (install if missing):
#   install.packages(c("MatchIt", "optmatch", "cobalt", "ggplot2"))

library(MatchIt)
library(cobalt)
library(ggplot2)

# ============================================================
# 1. Load LaLonde-PSID1 data (data/ lives one level above this script)
# ============================================================
data_path <- if (file.exists("../data/lalonde_psid.csv")) {
  "../data/lalonde_psid.csv"
} else if (file.exists("data/lalonde_psid.csv")) {
  "data/lalonde_psid.csv"
} else {
  "transcripts/code/data/lalonde_psid.csv"
}
df <- read.csv(data_path)
covars <- c("age", "educ", "black", "hisp", "marr", "nodegree", "re74", "re75")
form   <- as.formula(paste("treat ~", paste(covars, collapse = " + ")))

# ============================================================
# 2. Naive (unmatched) comparison
# ============================================================
cat("---- Naive difference in means (re78) ----\n")
print(t.test(re78 ~ treat, data = df))

# ============================================================
# 3. Logistic propensity score and overlap inspection
# ============================================================
ps_model    <- glm(form, data = df, family = binomial)
df$pscore   <- predict(ps_model, type = "response")
df$logit_ps <- predict(ps_model)         # logit scale: better for matching

ggplot(df, aes(x = pscore, fill = factor(treat))) +
  geom_density(alpha = 0.4) +
  labs(title = "Propensity score overlap", fill = "Treated")

# ============================================================
# 4. Optimal pair matching with a 0.2 SD caliper on logit propensity
# ============================================================
m1 <- matchit(form, data = df,
              method   = "optimal",
              distance = "glm",         # logistic propensity
              caliper  = 0.2)
cat("\n---- Optimal 1:1 PS matching ----\n")
print(summary(m1)$nn)

# ============================================================
# 5. Balance diagnostics: standardized differences + Love plot
# ============================================================
cat("\n---- Standardized differences before/after ----\n")
print(bal.tab(m1, m.threshold = 0.1))

love.plot(m1, threshold = 0.1, abs = TRUE,
          colors = c("#0065BD", "#E37222"),
          shapes = c("circle", "triangle"),
          title  = "Love plot: NSW matching")

# ============================================================
# 6. Multiple controls per treated (1:k matching)
# k=2 trades a bit of bias for ~25% variance reduction (Notes 6).
# ============================================================
m2 <- matchit(form, data = df,
              method   = "nearest",
              ratio    = 2,
              distance = "glm",
              caliper  = 0.2,
              replace  = FALSE)
cat("\n---- 1:2 nearest-neighbour matching ----\n")
print(summary(m2)$nn)

# ============================================================
# 7. Mahalanobis distance + propensity caliper hybrid
# Match locally on key prognostic vars, globally on propensity.
# ============================================================
m3 <- matchit(form, data = df,
              method       = "nearest",
              distance     = "glm",       # propensity for caliper
              mahvars      = ~ age + educ + re74 + re75,
              caliper      = 0.2)
cat("\n---- Mahalanobis + propensity caliper hybrid ----\n")
print(summary(m3)$nn)

# ============================================================
# 8. Inference: regression on the matched dataset
# ============================================================
matched <- match.data(m1)
fit <- lm(re78 ~ treat + age + educ + black + hisp + marr +
            nodegree + re74 + re75,
          data    = matched,
          weights = weights)
cat("\n---- Treatment effect on matched data ----\n")
print(coef(summary(fit))["treat", ])

# ============================================================
# 9. Doubly robust preview (full IPW + outcome regression)
#    Detailed coverage in Session 5.
# ============================================================
df$w <- ifelse(df$treat == 1, 1 / df$pscore, 1 / (1 - df$pscore))
fit_dr <- lm(re78 ~ treat + age + educ + black + hisp + marr +
               nodegree + re74 + re75,
             data    = df,
             weights = w)
cat("\n---- Doubly robust (IPW-weighted regression) ----\n")
print(coef(summary(fit_dr))["treat", ])

File: 26S/session4.py
# session4.py
# Session 4: Matching Methods & Propensity Scores
# End-to-end runnable example using the Dehejia-Wahba LaLonde-PSID1 sample
# (185 NSW treated + 2,490 PSID-1 observational controls). The severe
# imbalance between these two groups is what makes matching interesting --
# the original NSW experimental sample is already balanced by randomization.
# Demonstrates: PSM, nearest-neighbour matching, balance diagnostics, ATT.
#
# We use the `causalinference` package (Wasserman / Causal Inference textbook
# companion) — a more mature option than psmpy for an econ/stats course.
#
# Required packages:
#   pip install causalinference pandas numpy statsmodels matplotlib

from pathlib import Path

import numpy as np
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt

from causalinference import CausalModel

DATA_PATH = Path(__file__).resolve().parent.parent / "data" / "lalonde_psid.csv"
df = pd.read_csv(DATA_PATH).reset_index(drop=True)
covars = ["age", "educ", "black", "hisp", "marr", "nodegree", "re74", "re75"]

# ============================================================
# 1. Naive (unmatched) comparison
# ============================================================
y_t = df.loc[df.treat == 1, "re78"]
y_c = df.loc[df.treat == 0, "re78"]
print("---- Naive difference in means ----")
print(f"  Treated mean: {y_t.mean():.1f}")
print(f"  Control mean: {y_c.mean():.1f}")
print(f"  Diff:         {y_t.mean() - y_c.mean():+.1f}")

# ============================================================
# 2. Build a CausalModel object: (Y, D, X)
# ============================================================
Y = df["re78"].values
D = df["treat"].values
X = df[covars].values
cm = CausalModel(Y, D, X)

# Pre-matching covariate summary (Raw-diff and Norm-diff per covariate).
print("\n---- Pre-matching covariate summary ----")
print(cm.summary_stats)

# ============================================================
# 3. Estimate propensity score (logistic) and visualise overlap
# ============================================================
cm.est_propensity_s()           # stepwise logistic with quadratic terms
df["pscore"] = cm.propensity["fitted"]

fig, ax = plt.subplots(figsize=(6, 3))
df.loc[df.treat == 1, "pscore"].plot.kde(ax=ax, label="Treated")
df.loc[df.treat == 0, "pscore"].plot.kde(ax=ax, label="Control")
ax.set_title("Propensity score overlap")
ax.set_xlabel("p(T=1|X)"); ax.legend()

# ============================================================
# 4. Trim units outside the common-support region
#    (cuts treated and controls with extreme propensity scores).
# ============================================================
cm.cutoff = 0.1
cm.trim()
print(f"\n---- After trimming: N = {cm.raw_data['N']} ----")

# ============================================================
# 5. Nearest-neighbour matching on propensity score (1:1)
#    `est_via_matching` returns ATE / ATT / ATC with Abadie-Imbens SEs.
# ============================================================
cm.est_via_matching(matches=1, bias_adj=True)
print("\n---- Matching estimates (Abadie-Imbens bias-adjusted) ----")
print(cm.estimates)

# ============================================================
# 6. Balance diagnostics: standardized differences before vs. after
#    matching, plus a manual Love plot.
#
# `causalinference` does not expose a public matched-index attribute, so we
# reconstruct 1-NN propensity matches with sklearn. This gives us a long-
# format matched DataFrame, which we use for post-matching SMDs.
# ============================================================
from sklearn.neighbors import NearestNeighbors  # noqa: E402

def smd(t, c):
    return (t.mean() - c.mean()) / np.sqrt((t.var() + c.var()) / 2)

treated_idx = np.where(D == 1)[0]
control_idx = np.where(D == 0)[0]
nn = NearestNeighbors(n_neighbors=1).fit(
    df.loc[control_idx, "pscore"].to_numpy().reshape(-1, 1))
_, nn_idx = nn.kneighbors(
    df.loc[treated_idx, "pscore"].to_numpy().reshape(-1, 1))
matched_ctrl_idx = control_idx[nn_idx.flatten()]
matched_idx = np.concatenate([treated_idx, matched_ctrl_idx])
matched = df.iloc[matched_idx].copy()

print("\n---- Standardized mean differences: before vs. after matching ----")
print(f"  {'covariate':<10}  {'before':>8}  {'after':>8}")
smd_before, smd_after = {}, {}
for v in covars:
    b = smd(df.loc[df.treat == 1, v],     df.loc[df.treat == 0, v])
    a = smd(matched.loc[matched.treat == 1, v],
            matched.loc[matched.treat == 0, v])
    smd_before[v], smd_after[v] = b, a
    flag = "OK" if abs(a) < 0.1 else ("WARN" if abs(a) < 0.2 else "BAD")
    print(f"  {v:<10}  {b:+8.3f}  {a:+8.3f}  [{flag}]")

# ----- Manual Love plot (matplotlib) ---------------------------
fig, ax = plt.subplots(figsize=(6, 4))
ax.scatter([abs(v) for v in smd_before.values()], covars,
           marker="x", color="#0065BD", label="Before matching")
ax.scatter([abs(v) for v in smd_after.values()],  covars,
           marker="o", color="#E37222", label="After matching")
ax.axvline(0.1, ls="--", color="grey")
ax.set_xlabel("|Standardized mean difference|")
ax.set_title("Love plot: NSW matching")
ax.legend()
fig.tight_layout()

# ============================================================
# 7. Doubly robust preview (full data, IPW-weighted regression).
#    Detailed coverage in Session 5.
# ============================================================
df["w"] = np.where(df.treat == 1, 1 / df.pscore, 1 / (1 - df.pscore))
X_full  = sm.add_constant(df[["treat"] + covars])
fit_dr  = sm.WLS(df["re78"], X_full, weights=df["w"]).fit()
print(f"\n---- Doubly robust (IPW-weighted regression) ----")
print(f"  tau = {fit_dr.params['treat']:+8.1f}  (SE = {fit_dr.bse['treat']:.1f})")
